Add LDAP authentication

immae opened the pull request:
This PR implements LDAP authentication. The behavior is the expected standard behavior for LDAP authentication: a master LDAP account can search for the user that tries to log-in, then if found we try to bind as that user and if successful, the user is logged in.

I tried to add some tests, but I couldn’t find any existing tests for models and libs (the parts of the code that I touched)...

Fixes https://github.com/Chocobozzz/PeerTube/issues/371

ldidry commented on config/production.yaml.example:
```diff
@@ -33,6 +33,17 @@ redis:
+  base: dc=example,dc=com
+  mail_entry: "mail"
+  user_filter: "(|(email=%username%)(uid=%username%))"
+
```
You may want to add a setting to prevent connection failure when you use ldaps and the LDAP server has a self-signed certificate (for ex.).

I didn't use the same module than you to connect to LDAP, but on [ep_mypads](https://framagit.org/framasoft/Etherpad/ep_mypads/), I had to add this optional setting in the LDAP settings:

```
    "tlsOptions": {
        "rejectUnauthorized": false
    },
```

If you want to easily test your code against a self-signed certificated ldaps, you can use this docker image: https://github.com/rroemhild/docker-test-openldap

immae replied:
Good catch @ldidry I added that option.

I used inspiration from etherpad-lite to write this PR, so it’s definitely the same node module ldapjs that is used in both cases :)

(And thanks for pointing me to ep_mypads module, it seems great and I didn’t know about!)

Chocobozzz commented on config/default.yaml:
```diff
@@ -33,6 +33,9 @@ redis:
   db: 0
 
+ldap:
+  enable: false
```
`enabled` for consistency

And add all ldap keys please

Chocobozzz commented on config/production.yaml.example:
```diff
@@ -33,6 +33,18 @@ redis:
   db: 0
 
+ldap:
+  enable: true
```
`enabled`

Chocobozzz commented on config/production.yaml.example:
```diff
@@ -33,6 +33,18 @@ redis:
+ldap:
+  enable: true
+  # Disallow non-ldap users (it also disables root!)
+  ldap_only: true
```
I'm not sure it's a good idea. What if instead we have something like:

```
auth:
  local:
    enabled: boolean
  ldap:
    enabled: boolean
    ...
```

immae replied:
After some thought, I don’t know if it is a good idea to support both, apart from a transitionning period: supporting both local and "remote" (be it ldap or something else) user will most probably provoke clashes in usernames at some point.

So my thought would be to have ldap_only to true, and warn the user that ldap_only: false should be used only during transition to ldap

Chocobozzz replied:
Yes, it was my second concern:

"What do we do when someone registers on the instance with a username that already exists in LDAP, but not in our local database (because the user has not yet connected for example)"

When someone registers, we could check in LDAP if the username already exists. Because even if there is not much sense to have LDAP and local auth enabled, we'll have this kind of use case with other authentication methods.

Anyway, please use the `auth:` parent key. We can then throw errors/warnings at startup if we decide we don't support LDAP & local authentication.

Chocobozzz commented on server/initializers/constants.ts:
```diff
@@ -16,7 +16,7 @@ let config: IConfig = require('config')
 // ---------------------------------------------------------------------------
 
-const LAST_MIGRATION_VERSION = 340
+const LAST_MIGRATION_VERSION = 341
```
Use 345 so we can insert other migrations in patch versions

immae replied:
(changed for 375)

Chocobozzz commented on server/initializers/constants.ts:
```diff
@@ -177,6 +177,17 @@ const CONFIG = {
+    BIND_PASSWORD: config.has('ldap.bind_password') ? config.get<string>('ldap.bind_password') : null,
+    BASE: config.has('ldap.base') ? config.get<string>('ldap.base') : null,
+    MAIL_ENTRY: config.has('ldap.mail_entry') ? config.get<string>('ldap.mail_entry') : 'mail',
+    USER_FILTER: config.has('ldap.user_filter') ? config.get<string>('ldap.user_filter') : '(|(email=%username%)(uid=%username%))'
```
Prefer `null` instead of `'(|(email=%username%)(uid=%username%))'`

immae replied:
Ok

Chocobozzz commented on server/initializers/constants.ts:
```diff
@@ -177,6 +177,17 @@ const CONFIG = {
+    BIND_DN: config.has('ldap.bind_dn') ? config.get<string>('ldap.bind_dn') : null,
+    BIND_PASSWORD: config.has('ldap.bind_password') ? config.get<string>('ldap.bind_password') : null,
+    BASE: config.has('ldap.base') ? config.get<string>('ldap.base') : null,
+    MAIL_ENTRY: config.has('ldap.mail_entry') ? config.get<string>('ldap.mail_entry') : 'mail',
```
Prefer `null` instead of `'mail'`

immae replied:
`mail` is a standard field name for e-mail in ldap, so I think we should keep it

Chocobozzz commented on server/lib/oauth-model.ts:
```diff
@@ -66,7 +66,13 @@ function getRefreshToken (refreshToken: string) {
   logger.debug('Getting User (username/email: ' + usernameOrEmail + ', password: ******).')
 
-  const user = await UserModel.loadByUsernameOrEmail(usernameOrEmail)
+  let user
```
`let user: UserModel`

Chocobozzz commented on server/models/account/user.ts:
```diff
@@ -354,6 +361,91 @@ export class UserModel extends Model<UserModel> {
+      .catch(() => { return null })
+  }
+
+  static findLDAPUser (username: string) {
```
Put it in a `lib/ldap.ts` module, so you create the client only once

Chocobozzz commented on server/models/account/user.ts:
```diff
@@ -354,6 +361,91 @@ export class UserModel extends Model<UserModel> {
+  static findOrCreateLDAPUser (username: string) {
+    let userInfos
+
+    return Promise.resolve(UserModel.findLDAPUser(username))
```
Use async/await instead

Example: `const { dn ] = await UserModel.findLDAPUser(username)` etc

You can use try/catch, but please log the error

Chocobozzz commented on server/models/account/user.ts:
```diff
@@ -354,6 +361,91 @@ export class UserModel extends Model<UserModel> {
+        if (err) reject()
+        let entries = []
+        client.search(CONFIG.LDAP.BASE, opts, function (err, search) {
+          if (err) reject()
```
`reject(err)`

Chocobozzz commented on server/models/account/user.ts:
```diff
@@ -354,6 +361,91 @@ export class UserModel extends Model<UserModel> {
+
+    return new Promise(function (resolve, reject) {
+      client.bind(CONFIG.LDAP.BIND_DN, CONFIG.LDAP.BIND_PASSWORD, function (err) {
+        if (err) reject()
```
`reject(err)`

Chocobozzz commented on server/models/account/user.ts:
```diff
@@ -354,6 +361,91 @@ export class UserModel extends Model<UserModel> {
+            if (entries.length === 1) {
+              resolve(entries[0])
+            } else {
+              reject()
```
Reject an error please

Chocobozzz commented on server/models/account/user.ts:
```diff
@@ -354,6 +361,91 @@ export class UserModel extends Model<UserModel> {
+  }
+
+  static createLDAPUser (username: string, userInfos: {}) {
+    return Promise.resolve(userInfos)
```
Use async/await

You don't need `Promise.resolve`

Chocobozzz commented on server/models/account/user.ts:
```diff
@@ -354,6 +361,91 @@ export class UserModel extends Model<UserModel> {
+      .then((userInfos) => {
+        const userToCreate = new UserModel({
+          username: username,
+          password: 'SomeInvalidPassword',
```
`null` no?

immae replied:
password field cannot be null, and I didn’t want to remove this constraint as I don’t know the consequences in the rest of the code. If you think it’s safe, we can make it nullable

Chocobozzz replied:
Yep I think we can allow null on this column

Chocobozzz commented on server/models/account/user.ts:
```diff
@@ -354,6 +361,91 @@ export class UserModel extends Model<UserModel> {
+    return Promise.resolve(userInfos)
+      .then((userInfos) => {
+        const userToCreate = new UserModel({
+          username: username,
```
`username: username,` -> `username,`

immae commented:
> What if the user username or email is changed in LDAP? We can say we don't support username change, but we should support email change.

Right, we should update the e-mail when it changes on LDAP side

> What do we do when someone registers on the instance with a username that already exists in LDAP, but not in our local database (because the user has not yet connected for example)

It is created automatically

immae commented:
@Chocobozzz : I finally had the time to address your requests.

I didn’t manage to make the password field nullable : I changed the column in the database to make it nullable, changed in the UserModel too, and made the isUserPasswordValid validator more flexible, but it keeps failing. I may have missed something but I couldn’t find what.

I changed some of the Promise as required, but in lib/ldap.ts there are still some: the ldapjs library works with callbacks, and I don’t know how to turn that into an async/await process.

immae commented:
(about password field nullable: it fails in `lib/user.ts` `createUserAccountAndChannelAndPlaylist`, at  line `userToCreate.save(userOptions)`, regardless of `validate` value)

Chocobozzz commented:
> I didn’t manage to make the password field nullable : I changed the column in the database to make it nullable, changed in the UserModel too, and made the isUserPasswordValid validator more flexible, but it keeps failing. I may have missed something but I couldn’t find what.

No problem, I'll merge your PR in a separate branch (next week or the week after) and will update the code myself. 

Please could you add unit tests?

Chocobozzz commented on server/lib/oauth-model.ts:
```diff
@@ -75,7 +75,13 @@ function getRefreshToken (refreshToken: string) {
+  if (CONFIG.AUTH.LDAP.ENABLED) {
+    user = await UserModel.findOrCreateLDAPUser(usernameOrEmail)
+  }
+  if (!user && CONFIG.AUTH.LOCAL.ENABLED) {
```
I think we should put the local check first. Because users that were created before the LDAP activation should be able to log in with their "old" account.

battosai30 replied:
Usually it's not the good strategy : 
1) If LDAP is enabled, you first  check if login and password are ok in LDAP server.
2) If not, and if fallback login option is activated, you check in local app database if login/password are OK.

Usually LDAP server is the authentification security reference, so it's normal that's if activated it's checked first.

Chocobozzz commented:
> >  What do we do when someone registers on the instance with a username that already exists in LDAP, but not in our local database (because the user has not yet connected for example)

> It is created automatically

IMHO you should add a check in the registration controller/middleware, so users cannot register a username or email that already exists in LDAP. What do you think?

immae commented:
[IRC.log](https://github.com/Chocobozzz/PeerTube/files/3171667/IRC.log)

immae commented:
[IRC.log](https://github.com/Chocobozzz/PeerTube/files/3171672/IRC.log)

battosai30 commented:
Hi guys,

Any update on this feature ? I'm thinking about implement PeerTube in severall structure that I managed but I really need LDAP for my users as I already use other services (like RocketChat, Nextcloud etc ...) which needs authentification and already support LDAP and allowed my users to only have a single login set on all platforms.

Regards and thanks for your great job

Chocobozzz commented:
Hi,

I plan to improve the plugin system that would allow to register an `auth` method. Then, (I or @immae if he wants) will externalize this PR in a plugin.

The algo would be:
 * Every auth plugin would have a weight, defined by the administrator
 * On login, we check if the username exists in database. If yes, try to login against this user (it could be an external auth method)
 * If the user does not exist, try to login with every auth plugin ordered by `weight DESC`. On success with a particular auth plugin, create the user in the database with a column that specify the auth method (`LDAP`, `local` etc)

@immae What do you think?

immae commented:
@Chocobozzz yes that seems to be perfect :) Maybe some details would need to be adapted (for auth methods that require more than just a password), but the idea seems good

jadjay commented:
Thanks for this !
I hope it will be soon released...

immae commented:
@Chocobozzz : should we maybe close this stale PR? Please note that I’m  following the project (with a little delay, I’m currently at latest 2.X.X), so I’m always able to provide an up to date PR whenever the topic comes back on the table

immae commented:
NB: I didn’t follow the project closely enough to be able to tell wether you implemented that plugin system you talk about in above comment, if so please tell me and I can have a look at it...

Chocobozzz commented:
@immae Nope, it's a reminder for me. If all goes well, I should develop an LDAP plugin this year.

Chocobozzz commented:
Plugin done in https://framagit.org/framasoft/peertube/official-plugins/-/tree/master/peertube-plugin-auth-ldap

Plugin auth support WIP on https://github.com/Chocobozzz/PeerTube/tree/feature/auth-plugins

UnlimitedCookies commented:
Thank you so much for creating the highly requested ldap auth plugin! ❤️ 

I am still wondering: 
- How and when can I install this? (Because I couldn't find it in the plugin search yet)
- And will there be documentation (for general OpenLDAP usage)?
- Where can I find the auth weight setting then?

rigelk commented:
@UnlimitedCookies the plugin specifies compatibility only starting with the next release, as can be read in its `package.json`, so you will have to wait for version 2.2 :wink:

Chocobozzz commented:
> And will there be documentation (for general OpenLDAP usage)?
Where can I find the auth weight setting then?

I'm not sure to understand these questions

willrimmer commented:
Looking forward to LDAP support! One login to rule them all...

UnlimitedCookies commented:
@willrimmer And the amazing thing is, it's already implemented ;)

jdaviescoates commented:
@UnlimitedCookies yes, but how to actually implement/ use it? Needs documenting.

willrimmer commented:
Thats awesome! Is there guidelines on implementation?

July 23, 2020 6:49 PM, "UnlimitedCookies" <notifications@github.com (mailto:notifications@github.com?to=%22UnlimitedCookies%22%20<notifications@github.com>)> wrote:
	@willrimmer (http://url2781.famfam.us/ls/click?upn=eLZImTJF5xLTGfznWHzZtpmbFRbx0RofvM7Beoi2fQ12xmdQ-2FmozCUj987QfjUCc3931_U31jt5HgHBh7k3Grc5s4l7xuQERo1tl9uwqaRQ-2FWpkyKdwlIl44SYpn94CEvcAa1yKzGkaGD3jvkcCK4MiQYhKiXcNHIKEQY7nNDHNBSjQpQYyxq3D7NzgPRhkpSsvhtpYw01LZB9Mf4HC-2F-2FPfJHUPJpmh52jdCc8xpRbSkhQwmb7R-2FX-2BMNMZFaXrx-2FLomAZsQAIT9GGWfZ-2BJJ4kgDz-2BZbJ7dvzuhJmLOKkBtEKIC-2BGRtgCnNlDCGfTaG7GY-2BOpVd8jU7vh8VRkYnND0rN13YNv3Ab5GOML6Co-2FphaByHbk-3D And the amazing thing is, it's already implemented ;)

	—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub (http://url2781.famfam.us/ls/click?upn=eLZImTJF5xLTGfznWHzZtvSdfvTAdJ9H9UgV2vv1Tz0FLaxyiuIwbNENx9Hg1-2FiWYyp5iNjLCIQsYgv-2B4R9vpSdFPitQyd6t5UG-2BdEIbI1sfbXOAEscg1VU9UVaZjVmj4dSM_U31jt5HgHBh7k3Grc5s4l7xuQERo1tl9uwqaRQ-2FWpkyKdwlIl44SYpn94CEvcAa1yKzGkaGD3jvkcCK4MiQYhKiXcNHIKEQY7nNDHNBSjQpQYyxq3D7NzgPRhkpSsvhtE0f6WYltPW-2B2fyOaSu-2BdmTcRtxhbIXY7LfOLvvH8mTKsV73iiFUIB89ifQQKAcgpfkIxwEh-2BjV-2BTnw9GMPpliL17Zm-2FLdC-2BqM1lEdndZJTWSfzjDTO8lUQ-2FOvwGfhFcBodJFHy-2F5HGlixTylN3tbXqULvASsh0CLLFpVLxztN34-3D or unsubscribe (http://url2781.famfam.us/ls/click?upn=eLZImTJF5xLTGfznWHzZtgTq7SB76uB7Lqj7nJILYWSg4yH9kQS3Y-2FOE4VKio57-2FZ-2FLmgB050q5-2FbaL2zH303158Udz4b92xX3xkwR1rKeSqmjJabLXydS4m3qr87A50-2FwPskkjXO9B7cfxsdqiAeA-3D-3DDqdd_U31jt5HgHBh7k3Grc5s4l7xuQERo1tl9uwqaRQ-2FWpkyKdwlIl44SYpn94CEvcAa1yKzGkaGD3jvkcCK4MiQYhKiXcNHIKEQY7nNDHNBSjQpQYyxq3D7NzgPRhkpSsvht8hwApCFerm0GeKx5ApMSpgIjSMewwJFcaIO9XHJdWfCSNGxvZ20MORd1HdfY2yKkG4GyvwR6gDXubZm-2Fan9yysj-2BKPzTOWSQ-2F1UjAbeKwDBFeyuwOZn84F9-2FAqJEh6axuVyUhhxF0xtPfE50hqN0r63kyYpVI6Bt2cggQuuB3dE-3D

Chocobozzz commented:
@jdaviescoates https://docs.joinpeertube.org/#/admin-managing-users?id=external-auths-support
